### ISSUE TITLE
[bitnami/grafana-operator] CRDs GrafanaDatasource and GrafanaDashboard not being added to Grafana

### ISSUE THREAD

**Issue by ledroide** [bitnami/charts#24865]
### Name and Version

bitnami/grafana-operator 4.0.0

### What architecture are you using?

amd64

#### summary

Resources configured and deployed from CRDs GrafanaDatasource and GrafanaDashboard were not added to Grafana.

##### helm charts

* bitnami/grafana-operator 4.0.0
* bitnami/kube-prometheus 9.0.0
* bitnami/grafana-loki 3.0.1

##### kubernetes environment

* kubernetes v1.29.2 on premise
* cilium v1.15.2
* cri-o 1.29.1 ; crun 1.8.5

#### values for grafana-operator

```yaml
global:
  storageClass: local-path
operator:
  serviceAccount:
    automountServiceAccountToken: true
grafana:
  persistence:
    enabled: true
    size: 1Gi
    accessModes:
      - ReadWriteMany
  labels:
    app: grafana
    dashboards: grafana
```

#### how to reproduce

Here is the result of the deployed helm chart in the _monitoring_ namespace. Looks fine.

```console
$ kubectl get grafana,deploy,grafanadatasources,grafanadashboards,pod -L dashboards -L app.kubernetes.io/component -L app -l dashboards=grafana
NAME                                                               STAGE      STAGE STATUS   AGE   DASHBOARDS   COMPONENT   APP
grafana.grafana.integreatly.org/grafana-grafana-operator-grafana   complete   success        77m   grafana                  grafana

NAME                                                          READY   UP-TO-DATE   AVAILABLE   AGE    DASHBOARDS   COMPONENT   APP
deployment.apps/grafana-grafana-operator-grafana-deployment   1/1     1            1           7m7s   grafana      grafana     grafana

NAME                                                   NO MATCHING INSTANCES   LAST RESYNC   AGE     DASHBOARDS   COMPONENT   APP
grafanadatasource.grafana.integreatly.org/loki                                               5m45s   grafana      grafana     grafana
grafanadatasource.grafana.integreatly.org/prometheus                                         5m45s   grafana      grafana     grafana

NAME                                                        NO MATCHING INSTANCES   LAST RESYNC   AGE     DASHBOARDS   COMPONENT   APP
grafanadashboard.grafana.integreatly.org/k8s-views-global                                         5m45s   grafana      grafana     grafana

```

Here are available CRDs managed by the grafana operator. Still looks good to me :

```console
$ kubectl api-resources  --api-group grafana.integreatly.org
NAME                     SHORTNAMES   APIVERSION                        NAMESPACED   KIND
grafanaalertrulegroups                grafana.integreatly.org/v1beta1   true         GrafanaAlertRuleGroup
grafanadashboards                     grafana.integreatly.org/v1beta1   true         GrafanaDashboard
grafanadatasources                    grafana.integreatly.org/v1beta1   true         GrafanaDatasource
grafanafolders                        grafana.integreatly.org/v1beta1   true         GrafanaFolder
grafanas                              grafana.integreatly.org/v1beta1   true         Grafana
```

Here are _GrafanaDatasource_ manifests for Prometheus and Loki :

```yaml
apiVersion: grafana.integreatly.org/v1beta1
kind: GrafanaDatasource
metadata:
  labels:
    app: grafana
    app.kubernetes.io/component: grafana
    dashboards: grafana
  name: loki
  namespace: monitoring
spec:
  allowCrossNamespaceImport: true
  datasource:
    editable: true
    isDefault: false
    jsonData:
      timeInterval: 5s
      tlsSkipVerify: true
    name: Loki
    type: loki
    url: http://loki-grafana-loki-querier:3100
  instanceSelector:
    matchLabels:
      dashboards: grafana
```
```yaml
apiVersion: grafana.integreatly.org/v1beta1
kind: GrafanaDatasource
metadata:
  labels:
    app: grafana
    app.kubernetes.io/component: grafana
    dashboards: grafana
  name: prometheus
  namespace: monitoring
spec:
  allowCrossNamespaceImport: true
  datasource:
    editable: true
    isDefault: true
    jsonData:
      timeInterval: 5s
      tlsSkipVerify: true
    name: Prometheus
    type: prometheus
    url: http://prometheus-kube-prometheus-prometheus:9090
  instanceSelector:
    matchLabels:
      dashboards: grafana
```

Here is the _GrafanaDashboard_ manifest :

```yaml
apiVersion: grafana.integreatly.org/v1beta1
kind: GrafanaDashboard
metadata:
  annotations:
    github_source: https://github.com/dotdc/grafana-dashboards-kubernetes
  labels:
    app: grafana
    app.kubernetes.io/component: grafana
    dashboards: grafana
  name: k8s-views-global
  namespace: monitoring
spec:
  datasources:
  - datasourceName: Prometheus
    inputName: DS_PROMETHEUS
  instanceSelector:
    matchLabels:
      dashboards: grafana
  url: https://raw.githubusercontent.com/dotdc/grafana-dashboards-kubernetes/master/dashboards/k8s-views-global.json
```

#### reconcilation logs in grafana operator

We notice messages with "status: 400" "bad request data" "Internal error occurred" "context deadline exceeded" about the reconciliation loop.

```
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator github.com/grafana/grafana-operator/v5/controllers.(*GrafanaDatasourceReconciler).Reconcile
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator 2024-04-04T08:44:04Z INFO    GrafanaDatasourceReconciler     found matching Grafana instances for datasource {"controller": "grafanadatasource", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDatasource", "GrafanaDatasource": {"name":"prometheus","namespace":"monitoring"}, "namespace": "monitoring", "name": "prometheus", "reconcileID": "079b1ed5-27c2-4a50-afae-256226301da4", "count": 1}
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator 2024-04-04T08:44:04Z ERROR   GrafanaDatasourceReconciler     error reconciling datasource    {"controller": "grafanadatasource", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDatasource", "GrafanaDatasource": {"name":"prometheus","namespace":"monitoring"}, "namespace": "monitoring", "name": "prometheus", "reconcileID": "079b1ed5-27c2-4a50-afae-256226301da4", "datasource": "prometheus", "grafana": "grafana-grafana-operator-grafana", "error": "status: 400, body: {\"message\":\"bad request data\",\"traceID\":\"\"}"}
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator github.com/grafana/grafana-operator/v5/controllers.(*GrafanaDatasourceReconciler).Reconcile
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator 2024-04-04T08:44:04Z INFO    GrafanaDashboardReconciler      found matching Grafana instances for dashboard  {"controller": "grafanadashboard", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDashboard", "GrafanaDashboard": {"name":"k8s-views-global","namespace":"monitoring"}, "namespace": "monitoring", "name": "k8s-views-global", "reconcileID": "4bbbe744-3901-48f8-80d9-9ac41921057d", "count": 1}
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator 2024-04-04T08:44:04Z ERROR   GrafanaDashboardReconciler      error reconciling dashboard     {"controller": "grafanadashboard", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDashboard", "GrafanaDashboard": {"name":"k8s-views-global","namespace":"monitoring"}, "namespace": "monitoring", "name": "k8s-views-global", "reconcileID": "4bbbe744-3901-48f8-80d9-9ac41921057d", "dashboard": "k8s-views-global", "grafana": "grafana-grafana-operator-grafana", "error": "Internal error occurred: Get \"http://grafana-grafana-operator-grafana-service.monitoring:3000/api/folders?limit=1000&page=1\": context deadline exceeded"}
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator github.com/grafana/grafana-operator/v5/controllers.(*GrafanaDashboardReconciler).Reconcile
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator 2024-04-04T08:44:13Z INFO    GrafanaDatasourceReconciler     found matching Grafana instances for datasource {"controller": "grafanadatasource", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDatasource", "GrafanaDatasource": {"name":"loki","namespace":"monitoring"}, "namespace": "monitoring", "name": "loki", "reconcileID": "f9c7809b-71f5-40f3-a6a6-92fe97645e83", "count": 1}
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator 2024-04-04T08:44:13Z ERROR   GrafanaDatasourceReconciler     error reconciling datasource    {"controller": "grafanadatasource", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDatasource", "GrafanaDatasource": {"name":"loki","namespace":"monitoring"}, "namespace": "monitoring", "name": "loki", "reconcileID": "f9c7809b-71f5-40f3-a6a6-92fe97645e83", "datasource": "loki", "grafana": "grafana-grafana-operator-grafana", "error": "status: 400, body: {\"message\":\"bad request data\",\"traceID\":\"\"}"}
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator github.com/grafana/grafana-operator/v5/controllers.(*GrafanaDatasourceReconciler).Reconcile
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator 2024-04-04T08:44:14Z INFO    GrafanaDatasourceReconciler     found matching Grafana instances for datasource {"controller": "grafanadatasource", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDatasource", "GrafanaDatasource": {"name":"prometheus","namespace":"monitoring"}, "namespace": "monitoring", "name": "prometheus", "reconcileID": "7256d5db-31fe-43bc-8012-1c6ec34df7ab", "count": 1}
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator 2024-04-04T08:44:14Z ERROR   GrafanaDatasourceReconciler     error reconciling datasource    {"controller": "grafanadatasource", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDatasource", "GrafanaDatasource": {"name":"prometheus","namespace":"monitoring"}, "namespace": "monitoring", "name": "prometheus", "reconcileID": "7256d5db-31fe-43bc-8012-1c6ec34df7ab", "datasource": "prometheus", "grafana": "grafana-grafana-operator-grafana", "error": "status: 400, body: {\"message\":\"bad request data\",\"traceID\":\"\"}"}
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator github.com/grafana/grafana-operator/v5/controllers.(*GrafanaDatasourceReconciler).Reconcile
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator 2024-04-04T08:44:14Z INFO    GrafanaDashboardReconciler      found matching Grafana instances for dashboard  {"controller": "grafanadashboard", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDashboard", "GrafanaDashboard": {"name":"k8s-views-global","namespace":"monitoring"}, "namespace": "monitoring", "name": "k8s-views-global", "reconcileID": "c46fc530-3fbd-44a9-8891-c385ddd4a08b", "count": 1}
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator 2024-04-04T08:44:14Z ERROR   GrafanaDashboardReconciler      error reconciling dashboard     {"controller": "grafanadashboard", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDashboard", "GrafanaDashboard": {"name":"k8s-views-global","namespace":"monitoring"}, "namespace": "monitoring", "name": "k8s-views-global", "reconcileID": "c46fc530-3fbd-44a9-8891-c385ddd4a08b", "dashboard": "k8s-views-global", "grafana": "grafana-grafana-operator-grafana", "error": "Internal error occurred: Get \"http://grafana-grafana-operator-grafana-service.monitoring:3000/api/folders?limit=1000&page=1\": context deadline exceeded"}
grafana-grafana-operator-7c4c4b46bf-5zxtz grafana-operator github.com/grafana/grafana-operator/v5/controllers.(*GrafanaDashboardReconciler).Reconcile
```

#### what we have tried

* previous version of the chart - grafana-operator 3.12.3 -> same issue
* default value for `operator.serviceAccount.automountServiceAccountToken` -> no difference
* different labels -> no difference
* we also followed [this documentation](https://grafana.com/docs/grafana-cloud/developer-resources/infrastructure-as-code/grafana-operator/operator-dashboards-folders-datasources/)

#### additional info

* resource `GrafanaFolder` works fine, it creates the folder(s) in the Grafana Dashboards page

**Comment by ledroide** [comment]
Maybe similar to https://github.com/bitnami/charts/issues/24550 because of "context deadline exceeded" when applying a Grafana resource

**Comment by ledroide** [comment]
Fixed. This was a bug, already reported in https://github.com/grafana/grafana-operator/issues/1467 and https://github.com/grafana/grafana-operator/issues/1465 , both solved by https://github.com/grafana/grafana-operator/pull/1475 .

Image docker.io/bitnami/grafana-operator:5.8.1-debian-12-r1 fixes the issue, such as helm chart bitnami/grafana-operator 4.0.3 that embeds grafana-operator 5.8.1.

Should also fix #24550

**Comment by rafariossaa** [comment]
Hi, thanks for keeping track of it and reporting the fix.
